File: PyNet/ga2230Net/ListenerSocket.py
import socket
import struct

from .Packet import Packet
from .PacketBuilder import PacketBuilder


class ListenerSocket:
    """A socket handler for listening for data.
    Implements a server that listens for sender connections on all network
    interfaces.
    """
    def __init__(self, port: int, packet_builder: PacketBuilder) -> None:
        """
        :param port: the port to listen on
        :type port: int
        :param packet_builder: the packet builder to use
        :type packet_builder: PacketBuilder
        """
        self.server_socket: socket.socket = socket.socket(
            socket.AF_INET, socket.SOCK_DGRAM
        )

        self.server_socket.bind(("0.0.0.0", port))

        self.packet_builder = packet_builder

    def get_packet() -> Packet:
        """Gets a single packet sent from a sender.
        the packet will be wrapped around a Packet object.
        If there isnt an available packet on buffer it will wait for one.

        :return: a single packet sent
        :rtype: Packet
        """

        raw_packet = self.server_socket.recvfrom(1024)[0]

        header_length = int.from_bytes(raw_packet[0:2], "big")

        # struct strings are all chars + empty byte
        raw_header = raw_packet[2: 2 + header_length]

        header: str = (
            struct.unpack(f">{header_length}s", raw_header)[0]
        ).decode("utf-8")

        body_start = 2 + header_length
        return self.packet_builder.build_from_raw(
            header,
            raw_packet[body_start: body_start +
                       self.packet_builder.size_of(header)]
        )

File: RIONet/ga2230net/src/main/java/org/ga2230net/ListenerSocket.java
package org.ga2230net;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;

public class ListenerSocket {
    private final DatagramSocket datagramSocket;
    private final PacketBuilder builder;
    private final DatagramPacket rawPacket = new DatagramPacket(new byte[2048], 2048);
    public ListenerSocket(int port, PacketBuilder builder) throws IOException {
        this.builder = builder;

        datagramSocket = new DatagramSocket(port);
    }

    public Packet getPacket() throws IOException, SocketHandlerException {
        datagramSocket.receive(rawPacket);
        int headerLength = rawPacket.getData()[0] + rawPacket.getData()[1];
        byte[] rawHeader = new byte[headerLength];
        System.arraycopy(rawPacket.getData(), 2, rawHeader, 0, headerLength);
        String header = new String(rawHeader);
        byte[] rawBody = new byte[builder.sizeOf(header)];
        System.arraycopy(rawPacket.getData(), 2 + headerLength, rawBody, 0, builder.sizeOf(header));
        return builder.buildFromRaw(header, rawBody);
    }
}
